plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/client/GumbyClient.kt

```diff
@@ -101,48 +109,60 @@ class GumbyClient(private val project: Project) {
             throw e // pass along error to callee
         } finally {
-            CodetransformTelemetry.logApiLatency(
-                codeTransformApiNames = apiName,
-                codeTransformSessionId = CodeTransformTelemetryState.instance.getSessionId(),
-                codeTransformRunTimeLatency = calculateTotalLatency(startTime, Instant.now()),
+            telemetry.logApiLatency(
+                apiName,
+                startTime,
                 codeTransformUploadId = uploadId,
                 codeTransformJobId = jobId,
-                codeTransformRequestId = result?.responseMetadata()?.requestId(),
+                codeTransformRequestId = result?.responseMetadata()?.requestId()
             )
         }
     }
 
     suspend fun downloadExportResultArchive(jobId: JobId): MutableList<ByteArray> = amazonQStreamingClient.exportResultArchive(
         jobId.id,
         ExportIntent.TRANSFORMATION,
-        {
-                e ->
+        { e ->
             LOG.error(e) { "${CodeTransformApiNames.ExportResultArchive} failed: ${e.message}" }
-            CodetransformTelemetry.logApiError(
-                codeTransformApiNames = CodeTransformApiNames.ExportResultArchive,
-                codeTransformSessionId = CodeTransformTelemetryState.instance.getSessionId(),
-                codeTransformApiErrorMessage = e.message.toString(),
-                codeTransformJobId = jobId.id,
-            )
+            telemetry.apiError(e.localizedMessage, CodeTransformApiNames.ExportResultArchive, jobId.id)
         },
         { startTime ->
-            CodetransformTelemetry.logApiLatency(
-                codeTransformApiNames = CodeTransformApiNames.ExportResultArchive,
-                codeTransformSessionId = CodeTransformTelemetryState.instance.getSessionId(),
-                codeTransformRunTimeLatency = calculateTotalLatency(startTime, Instant.now()),
-                codeTransformJobId = jobId.id,
-            )
+            telemetry.logApiLatency(CodeTransformApiNames.ExportResultArchive, startTime, codeTransformJobId = jobId.id)
         }
     )
 
+    /*
+     * Adapted from [CodeWhispererCodeScanSession]
+     */
+    fun uploadArtifactToS3(url: String, fileToUpload: File, checksum: String, kmsArn: String, shouldStop: () -> Boolean) {
+        HttpRequests.put(url, APPLICATION_ZIP).userAgent(AwsClientManager.userAgent).tuner {
+            it.setRequestProperty(CONTENT_SHA256, checksum)
+            if (kmsArn.isNotEmpty()) {
+                it.setRequestProperty(SERVER_SIDE_ENCRYPTION, AWS_KMS)
+                it.setRequestProperty(SERVER_SIDE_ENCRYPTION_AWS_KMS_KEY_ID, kmsArn)
+            }
+        }
+            .connect { request -> // default connect timeout is 10s
+                val connection = request.connection as HttpURLConnection
+                connection.setFixedLengthStreamingMode(fileToUpload.length())
+                fileToUpload.inputStream().use { inputStream ->
```

Comment:
i know this is a move, but could you add some context on why we need to upload it like this if that's handy?

Reply:
Well, two things here:

1. The initial implementation was based on CodeWhisperers implementation
2. Since our uploads are much larger we decided to make it interruptable at any point when users click stop transform. -> Avoids wasting customer time waiting for upload to finish before being able to stop the job. This led to some modifications of 1.

I think we should replace this with the `ApacheHttpClient` instead, we are seeing some upload issues in some cases, and perhaps this would fix it. Doing some internal tests with it atm. However that would be another pr.